File: ruby/component/tcpserver.rb
# Created by Satoshi Nakagawa.
# You can redistribute it and/or modify it under the Ruby's license or the GPL2.

class TcpServer < NSObject
  attr_accessor :delegate, :port
  attr_reader :clients
  
  def initialize
    @port = 0
    @clients = []
  end
  
  def open
    close if @sock
    @buf = ''
    @sock = AsyncSocket.alloc.initWithDelegate(self)
    @active = @sock.acceptOnPort_error(@port, nil) != 0
    @sock = nil unless @active
    @active
  end
  
  def close
    return unless @sock
    @sock.disconnect
    @sock.release
    @sock = nil
    @active = false
  end
  
  def active?
    @active
  end
  
  def close_client(client)
    client.close
    @clients.delete(client)
  end
  
  def close_all_clients
    @clients.each {|c| c.close }
    @clients.clear
  end
  
  def onSocket_didAcceptNewSocket(sock, conn)
    c = TcpClient.alloc.init
    c.init_with_existing_connection(conn)
    c.delegate = self
    @clients << c
    @delegate.tcpserver_on_accept(self, c) if @delegate
  end
  
  
  def tcpClientDidConnect(sender)
    @delegate.tcpserver_on_connect(self, sender) if @delegate
  end
  
  def tcpClientDidDisconnect(sender)
    @delegate.tcpserver_on_disconnect(self, sender) if @delegate
    close_client(sender)
  end
  
  def tcpClient_error(sender, err)
    @delegate.tcpserver_on_error(self, sender, err) if @delegate
  end
  
  def tcpClientDidReceiveData(sender)
    @delegate.tcpserver_on_read(self, sender) if @delegate
  end
  
  def tcpClientDidSendData(sender)
    @delegate.tcpserver_on_write(self, sender) if @delegate
  end
end

File: Classes/Library/TCPServer.m
// LimeChat is copyrighted free software by Satoshi Nakagawa <psychs AT limechat DOT net>.
// You can redistribute it and/or modify it under the terms of the GPL version 2 (see the file GPL.txt).

#import "TCPServer.h"
#import "AsyncSocket.h"


@implementation TCPServer
{
    AsyncSocket* _conn;
    NSMutableArray *_clients;
}

- (id)init
{
    self = [super init];
    if (self) {
        _clients = [NSMutableArray new];
    }
    return self;
}

- (void)dealloc
{
    [_conn disconnect];
}

- (BOOL)open
{
    if (_conn) {
        [self close];
    }

    _conn = [[AsyncSocket alloc] initWithDelegate:self];
    _isActive = [_conn acceptOnPort:_port error:NULL];
    if (!_isActive) {
        [self close];
    }
    return _isActive;
}

- (void)close
{
    [_conn disconnect];
    _conn = nil;
    _isActive = NO;
}

- (void)closeClient:(TCPClient*)client
{
    [client close];
    [_clients removeObjectIdenticalTo:client];
}

- (void)closeAllClients
{
    for (TCPClient* c in _clients) {
        [c close];
    }
    [_clients removeAllObjects];
}

- (void)onSocket:(AsyncSocket*)sock didAcceptNewSocket:(AsyncSocket*)newSocket
{
    TCPClient* c = [[TCPClient alloc] initWithExistingConnection:newSocket];
    c.delegate = self;
    [_clients addObject:c];

    if ([_delegate respondsToSelector:@selector(tcpServer:didAccept:)]) {
        [_delegate tcpServer:self didAccept:c];
    }
}

- (void)tcpClientDidConnect:(TCPClient*)sender
{
    if ([_delegate respondsToSelector:@selector(tcpServer:didConnect:)]) {
        [_delegate tcpServer:self didConnect:sender];
    }
}

- (void)tcpClientDidDisconnect:(TCPClient*)sender
{
    if ([_delegate respondsToSelector:@selector(tcpServer:didDisconnect:)]) {
        [_delegate tcpServer:self didDisconnect:sender];
    }
    [self closeClient:sender];
}

- (void)tcpClient:(TCPClient*)sender error:(NSString*)error
{
    if ([_delegate respondsToSelector:@selector(tcpServer:client:error:)]) {
        [_delegate tcpServer:self client:sender error:error];
    }
}

- (void)tcpClientDidReceiveData:(TCPClient*)sender
{
    if ([_delegate respondsToSelector:@selector(tcpServer:didReceiveData:)]) {
        [_delegate tcpServer:self didReceiveData:sender];
    }
}

- (void)tcpClientDidSendData:(TCPClient*)sender
{
    if ([_delegate respondsToSelector:@selector(tcpServer:didSendData:)]) {
        [_delegate tcpServer:self didSendData:sender];
    }
}

@end
